### PR TITLE
Enable new FP16 and support mixed precision by MO

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Enable new FP16 IR generation that was disabled in #8650*
 - *Disable converting to FP32 by default in MO which allows mixed-precision IR*
 - *Replace usage of np.array to mo_array which forces fp32 dtype if no dtype was passed in MO*

### Tickets:
 - *66048*

### CODE REVIEW

**Comment by Maxim-Doronin** [comment]
@AntonDudchenko please take a closer look. It may have a critical impact on vpu functionality

**Comment by mvafin** [comment]
Because of replacements on MO side this PR have a lot of changes. To better review it, please look into separate commits.

File: model-optimizer/mo/front/common/partial_infer/utils.py [hunk 1]
```diff
@@ -169,12 +169,19 @@ def int64_array(value: Union[Iterable[Union[float, int]], float, int]) -> np.nda
...
+    return float32_array(value)
 
 
-def float32_array(value: Union[Iterable[Union[float, int]], float, int]) -> np.ndarray:
-    return np.array(value, dtype=np.float32)
+def mo_array(value: Union[Iterable[Union[float, int]], float, int], dtype=None) -> np.ndarray:
+    x = np.array(value, dtype=dtype)
+    if not isinstance(value, np.ndarray) and x.dtype == np.float64 and dtype != np.float64:
+        x = x.astype(np.float32)
+    return x
```
Comment (jane-intel) [inline model-optimizer/mo/front/common/partial_infer/utils.py:184]: Would be great to have unit tests for that

File: model-optimizer/unit_tests/extensions/back/ChangeOutputTypeAttributes_test.py [hunk 2]
```diff
@@ -48,19 +48,19 @@ def test_range_out_of_fp16_min(self):
         self.assertRaises(Error, ChangeOutputTypeAttributes().find_and_replace_pattern, graph)
 
     def test_cast_correct_case(self):
-        input_data = np.array([0, 1000, 4, 9, 0])
+        input_data = mo_array([0, 1000, 4, 9, 0])
```
Comment (jane-intel) [inline model-optimizer/unit_tests/extensions/back/ChangeOutputTypeAttributes_test.py:52]: I don't see these unit-test changes as necessary. more-over this could break previous test logic. mo_array is for the code and for conscious usage in tests. Please revert unit-tests/ changes

Comment (mvafin) [reply model-optimizer/unit_tests/extensions/back/ChangeOutputTypeAttributes_test.py:52]: I will try, but I think that will result in a lot of manual changes in unit_tests, because all unit_tests use bare np.array without thinking about types.

File: model-optimizer/extensions/middle/EltwiseChecker.py [hunk 3]
```diff
@@ -63,7 +63,7 @@ def mark_eltwise_node(self, node, feature_channel=None):
         feature_dim = min(1, tensor_shape.size - 1) if node.graph.graph['layout'] == 'NCHW' else -1
         if feature_channel is not None:
             feature_dim = feature_channel
-        ones = np.ones(len(tensor_shape))
+        ones = np.ones(len(tensor_shape), dtype=np.float32)
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/EltwiseChecker.py:66]: this is all about shapes. why do we make them float? can we make helper for ones and zeros with the same logic as mo_array? mo_zeros, mo_ones. Also -- only in the code, not in the tests.

Comment (jane-intel) [reply model-optimizer/extensions/middle/EltwiseChecker.py:66]: Agreed to abandon the idea of mo_zeros and mo_ones but check that all the types are valid -- especially for the integer cases

**Comment by ArtemySkrebkov-intel** [comment]
@olegmorenov 

Can you please generate a package of IR using this PR for us to test?

**Review by jane-intel — COMMENTED** [review]
Reviewed. Please do not force push any more
Have couple of questions and nitpick asks. Please answer and then I will approve from my side

File: tools/mo/openvino/tools/mo/middle/StridedSliceReplacer.py [hunk 4]
```diff
@@ -49,8 +49,8 @@ def find_and_replace_pattern(self, graph: Graph):
             if np.prod(input_shape) != np.prod(output_shape):
                 continue
 
-            shrink_axis_mask = node.soft_get('shrink_axis_mask', np.zeros(len(input_shape)))
-            new_axis_mask = node.soft_get('new_axis_mask', np.zeros(len(input_shape)))
+            shrink_axis_mask = node.soft_get('shrink_axis_mask', np.zeros(len(input_shape), dtype=np.float32))
+            new_axis_mask = node.soft_get('new_axis_mask', np.zeros(len(input_shape), dtype=np.float32))
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/middle/StridedSliceReplacer.py:53]: Changing type from fp64 to fp32 here is safe, are you sure that changing to bool wont result in some weird failure somewhere?

File: tools/mo/openvino/tools/mo/back/compress_quantized_weights.py [hunk 5]
```diff
@@ -103,6 +105,27 @@ def pattern(self):
             ]
         )
 
+    def pattern2(self):
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/back/compress_quantized_weights.py:108]: POT return deleted Converts BEFORE calling this transformation

File: tools/mo/openvino/tools/mo/ops/convolution.py [hunk 6]
```diff
@@ -287,3 +290,26 @@ def reverse_infer(node: Node):
                     break
             if shape is not None:
                 node.in_port(0).data.set_shape(shape)
+
+    @staticmethod
+    def type_infer(node):
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/ops/convolution.py:295]: This code doesn't work in any known to me model, but I suspect on broader model scope it can be working and it is relatively safe to change type of conv constant.

File: tools/mo/openvino/tools/mo/ops/convolution.py [hunk 7]
```diff
@@ -287,3 +290,26 @@ def reverse_infer(node: Node):
...
+        # in case of input values data type mismatch we try to change the type of the constant to match the type of
+        # input at index 0.
+        if in_type_1 in [np.float16, np.float32, np.float64] and in_type_0 != in_type_1 and in_node_1.op == 'Const':
+            in_node_1 = node.in_port(1).get_source().node
+            log.error("Changing Const node '{}' data type from {} to {} for Convolution operation".format(
+                in_node_1.soft_get('name', in_node_1.id), in_type_1, in_type_0),
+                extra={'is_warning': True})
+            convert_const_node_value_type(in_node_1, in_type_0)
+        if node.is_in_port_connected(2):
+            in_type_2 = node.in_port(1).get_data_type()
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/ops/convolution.py:308]: Agree, this can be deleted

**Review by KodiaqQ — APPROVED** [review]

File: tools/mo/openvino/tools/mo/back/compress_quantized_weights.py [hunk 8]
```diff
@@ -209,9 +232,14 @@ def dequantize_data(fake_quantize: Node, dst_type: type, quantized_type: type) -
...
-        if np.issubdtype(dst_type, np.floating):
-            dst_type = data_type_str_to_np(graph.graph['cmd_params'].data_type)
+        if 'convert' in match:
+            dst_type = match['convert'].dst_type
+            match['convert']['stop_value_propagation'] = False
+            Cast.infer(match['convert'])
+        else:
+            dst_type = match['const'].value.dtype
+            if np.issubdtype(dst_type, np.floating):
+                dst_type = data_type_str_to_np(graph.graph['cmd_params'].data_type)
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/back/compress_quantized_weights.py:242]: yes, it will be fp32 always, will replace

**Review by AntonDudchenko — APPROVED** [review]

File: tools/mo/unit_tests/extensions/back/compress_quantized_weights_test.py [hunk 9]
```diff
@@ -173,6 +214,33 @@ def test_data_type(self, model_dtype, original, transformed=None):
...
+            *connect('sub:0', '0:mul'),
+            *connect('scale:0', '1:mul'),
+            *connect('mul:0', 'output'),
+        ], nodes_with_edges_only=True)
+        (flag, resp) = compare_graphs(graph, graph_ref, 'output', check_op_attrs=True)
+        self.assertTrue(flag, resp)
+
 
 @generator
 class AccuracyCheckFP32Test(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/unit_tests/extensions/back/compress_quantized_weights_test.py:246]: As I understand this test it is checking fp32 case, which is basically the main case after new fp16 introduced

**Review by jane-intel — APPROVED** [review]
Good job 👍🏼

File: tools/mo/openvino/tools/mo/ops/convolution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jane-intel) [reply tools/mo/openvino/tools/mo/ops/convolution.py:295]: I disagree to have dead code here with basically no tests and no known use cases. We had problems with `type_infer` due to inconsistent types in constants which you addressed via `mo_array` everywhere. Lets get rid of it here and in the matmul too

File: tools/mo/unit_tests/extensions/back/compress_quantized_weights_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jane-intel) [reply tools/mo/unit_tests/extensions/back/compress_quantized_weights_test.py:246]: But we have no tests for this particular case)

File: tools/mo/openvino/tools/mo/middle/StridedSliceReplacer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jane-intel) [reply tools/mo/openvino/tools/mo/middle/StridedSliceReplacer.py:53]: I am sure

File: tools/mo/openvino/tools/mo/ops/convolution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/ops/convolution.py:295]: this is not a dead code, but a safety bag. It may work in unexpected cases

File: tools/mo/openvino/tools/mo/front/caffe/extractors/utils.py [hunk 10]
```diff
@@ -39,7 +40,9 @@ def embed_input(attrs: dict, port: int, name: str, value: np.array, bin_name: st
 
     # memory safe value conversion to numpy;
     # previously we used `np.array(value)` and it was greedy for memory on caffe models especially
-    val = np.ndarray(shape=(len(value),))
+    # previously we always created float64 np.ndarray, now we force float32, we can't get data type from "value" for
+    # Caffe, because it comes as float64 from protobuf
+    val = np.ndarray(shape=(len(value),), dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/caffe/extractors/utils.py:45]: My local experiments show that this change makes reading caffe models quicker slightly, about 10% better

**Review by jane-intel — APPROVED** [review]

File: tools/mo/openvino/tools/mo/front/caffe/extractors/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jane-intel) [reply tools/mo/openvino/tools/mo/front/caffe/extractors/utils.py:45]: this is only one aspect -- memory hungry models are more dangerous. my previous comment was about memory mostly.

**Comment by rkazants** [comment]
Please fix:
```
[ ERROR ]  Traceback (most recent call last):
  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/lib/python3.6/site-packages/openvino/tools/mo/utils/class_registration.py", line 278, in apply_transform
    for_graph_and_each_sub_graph_recursively(graph, replacer.find_and_replace_pattern)
  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/lib/python3.6/site-packages/openvino/tools/mo/middle/pattern_match.py", line 46, in for_graph_and_each_sub_graph_recursively
    func(graph)
  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/lib/python3.6/site-packages/openvino/tools/mo/front/tf/WhileNormalize.py", line 23, in find_and_replace_pattern
    self.normalize_loop_node(graph, node)
  File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-e2e/b/tmp/e2e_test_venv/lib/python3.6/site-packages/openvino/tools/mo/front/tf/WhileNormalize.py", line 36, in normalize_loop_node
    'value': mo_array(True, dtype=np.bool)}).create_node()
NameError: name 'mo_array' is not defined
```

File: tools/mo/openvino/tools/mo/front/onnx/hard_sigmoid_ext.py [hunk 11]
```diff
@@ -18,8 +17,8 @@ def replace_op(self, graph: Graph, node: Node):
         alpha = onnx_attr(node, 'alpha', 'f', default=0.2)
         beta = onnx_attr(node, 'beta', 'f', default=0.5)
 
-        hard_sigmoid = create_op_with_const_inputs(graph, HardSigmoid, {1: np.array(alpha), 2: np.array(beta)},
-                                                   {'name': node.name + '/HardSigmoid_'})
+        hard_sigmoid = create_op_with_const_inputs(graph, HardSigmoid, {1: mo_array(alpha), 2: mo_array(beta)},
+                                                   {'name': node.name + '/HardSiigmoid_'})
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/onnx/hard_sigmoid_ext.py:21]: please fix typo

File: tools/mo/openvino/tools/mo/front/caffe/scale_ext.py [hunk 12]
```diff
@@ -25,16 +23,16 @@ def extract(cls, node):
         
         if model is None and len(pb.bottom) == 1:
             # default weights and biases for scale layer if the caffemodel file doesn't contain them
-            model = NamedAttrsClass({'blobs': np.array([NamedAttrsClass({'data': np.array([1])}),
-                                                 NamedAttrsClass({'data': np.array([0])})])})
+            model = NamedAttrsClass({'blobs': mo_array([NamedAttrsClass({'data': mo_array([1])}),
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/caffe/scale_ext.py:26]: `mo_array` is not lucky name. We have `int64_array` and why is it called `mo_array` for int32. Please revert this change

**Review by rkazants — CHANGES_REQUESTED** [review]
Please fix failure. Also, I need to review these changes one more time.

File: tools/mo/openvino/tools/mo/front/caffe/scale_ext.py (hunk 12 above)
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/caffe/scale_ext.py:26]: mo_array has a special purpose. It acts as np.array, but if no dtype provided and resulting dtype is fp64 it forces fp32 dtype. The name was agreed with @jane-intel

File: src/inference/src/cnn_network_ngraph_impl.cpp [hunk 13]
```diff
@@ -441,6 +442,7 @@ void CNNNetworkNGraphImpl::reshape(const std::map<std::string, ngraph::PartialSh
                 manager.register_pass<::ngraph::pass::ConvertMulticlassNmsToMulticlassNmsIE>();
                 manager.register_pass<::ngraph::pass::ConvertMatrixNmsToMatrixNmsIE>();
                 manager.register_pass<::ngraph::pass::DisableConvertConstantFoldingOnConstPath>();
+                manager.register_pass<::ov::pass::DisableDecompressionConvertConstantFolding>();
```
Comment (rkazants) [inline src/inference/src/cnn_network_ngraph_impl.cpp:445]: I though for API 1.0 we should allow folding of compression `Convert` nodes, shouldn't we? Please clarify this

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 14]
```diff
@@ -356,6 +358,8 @@ static void TransformationUpToCPUSpecificOpSet(std::shared_ptr<ngraph::Function>
             });
 
     // List of enabled/disabled transformations
+    pass_config->disable<ov::pass::DisableDecompressionConvertConstantFolding>();
+    pass_config->disable<ov::pass::ConvertCompressedOnlyToLegacy>();
```
Comment (rkazants) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:362]: Why do we have two transformations `DisableDecompressionConvertConstantFolding` and 
                                                               `EnableDecompressionConvertConstantFolding`? And why don't we implement just one. Now the passes for different plugins create confusion because on the hand you use `DisableDecompressionConvertConstantFolding` and on the other hand `ConvertCompressedOnlyToLegacy` registers `EnableDecompressionConvertConstantFolding`. It really over-complicates understanding the logic

Comment (rkazants) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:362]: Why do we need to have two separate passes `ConvertCompressedOnlyToLegacy` and `DisableDecompressionConvertConstantFolding`? Why is not sufficient to manage just `ConvertCompressedOnlyToLegacy` overall?

File: src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 15]
```diff
@@ -112,7 +112,7 @@ bool ngraph::pass::CommonOptimizations::run_on_model(const std::shared_ptr<ngrap
     // before CommonOptimization pipeline execution
     manager.register_pass<ngraph::pass::MOCTransformations>(true, false);
 
-    manager.register_pass<ov::pass::ConvertCompressedOnlyToLegacy, false>();
+    manager.register_pass<ov::pass::ConvertCompressedOnlyToLegacy>();
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp:115]: I think it makes sense to have a comment that in majority of plugins IRv11 in FP16 is not supported but CPU and GNA plugins. So we have to convert IRv11 in FP16 to IRv10

File: src/plugins/intel_gna/gna_plugin.cpp [hunk 16]
```diff
@@ -738,6 +740,8 @@ void GNAPlugin::LoadNetwork(CNNNetwork & _network) {
         // UnrollTI should be the last transformation in the transformation pipeline
         manager.register_pass<ngraph::pass::UnrollTensorIterator>();
         const auto& pass_config = manager.get_pass_config();
+        pass_config->disable<ov::pass::ConvertCompressedOnlyToLegacy>();
+        pass_config->disable<ov::pass::DisableDecompressionConvertConstantFolding>();
```
Comment (rkazants) [inline src/plugins/intel_gna/gna_plugin.cpp:744]: the same comment

File: tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py [hunk 17]
```diff
@@ -179,12 +179,19 @@ def int64_array(value: Union[Iterable[Union[float, int]], float, int]) -> np.nda
...
 def float_array(value: Union[Iterable[Union[float, int]], float, int]) -> np.ndarray:
-    return np.array(value, dtype=np.float64)
+    return float32_array(value)
 
 
-def float32_array(value: Union[Iterable[Union[float, int]], float, int]) -> np.ndarray:
-    return np.array(value, dtype=np.float32)
+def mo_array(value: Union[Iterable[Union[float, int]], float, int], dtype=None) -> np.ndarray:
+    x = np.array(value, dtype=dtype)
+    if not isinstance(value, np.ndarray) and x.dtype == np.float64 and dtype != np.float64:
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:192]: not clear the logic here because `x.dtype == dtype` after line #191 and in the mean time there is expression `x.dtype == np.float64 and dtype != np.float64` how can it be? If it relates to `dtype==None` case, I propose to use `dtype is None`.

Also, please add description for this function. Not clear how it is used. What is a case when `not isinstance(value, np.ndarray)`?

File: tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py [hunk 18]
```diff
@@ -179,12 +179,19 @@ def int64_array(value: Union[Iterable[Union[float, int]], float, int]) -> np.nda
...
+        x = x.astype(np.float32)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:193]: what if in some code we have a list `x = [1,2,3]` and I call `np.array(x)`. So the result will be np.array of np.int32
In your case, if we substitute this with `mo_array(x)`, the result will be np.array of np.float32.

This function looks dangerous to me

File: tools/mo/openvino/tools/mo/front/caffe/scale_ext.py (hunk 12 above)
Comment (rkazants) [reply tools/mo/openvino/tools/mo/front/caffe/scale_ext.py:26]: I am fine to have this function but I don't support to change the code overall. Please revert this change and use `mo_array` for your particular task only

Comment (rkazants) [reply tools/mo/openvino/tools/mo/front/caffe/scale_ext.py:26]: what if in some code we have a list x = [1,2,3] and I call np.array(x). So the result will be np.array of np.int32
In your case, if we substitute this with mo_array(x), the result will be np.array of np.float32.

This function looks dangerous to me

File: tools/mo/openvino/tools/mo/front/common/layout.py [hunk 19]
```diff
@@ -22,7 +23,7 @@ def convert_shape(shape: np.array, permute: np.array):
     result = [0, 0, 0, 0]
     for ind, perm_ind in enumerate(permute):
         result[ind] = shape[perm_ind]
-    return np.array(result)
+    return mo_array(result)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/common/layout.py:26]: please revert

**Review by rkazants — CHANGES_REQUESTED** [review]
Disagree with these changes regarding "Replace usage of np.array to mo_array which forces fp32 dtype if no dtype was passed in MO". Please revert

File: tools/mo/openvino/tools/mo/front/caffe/scale_ext.py (hunk 12 above)
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/caffe/scale_ext.py:26]: > what if in some code we have a list x = [1,2,3] and I call np.array(x). So the result will be np.array of np.int32 In your case, if we substitute this with mo_array(x), the result will be np.array of np.float32.
> 
> This function looks dangerous to me

If result of np.array is int32 the result of mo_array will be int32 too. Only in one case it will be different. If result of np.array is fp64 and type wasn't provided, which means fp64 wasn't expected.

File: tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py (hunk 17 above)
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:192]: On line 191 if dtype is None the resulting dtype may be unexpected, so if it happened to become fp64 we force it to fp32. `not isinstance(value, np.ndarray)` is for the cases when we get actual np.ndarray in this function and we don't need to change the precision of it.

File: tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py (hunk 18 above)
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:193]: if np.array(x) return int32 mo_array will return int32 too

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 14 above)
Comment (mvafin) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:362]: These two passes cant be merged. `DisableDecompressionConvertConstantFolding` only disables constant folding for fp16 compression `Converts` otherwise first `ConstantFolding` pass will remove all `Converts` and upgrade all constants from fp16 to fp32 and it runs as early as possible. `ConvertCompressedOnlyToLegac` on the other hand runs after `ConstantFolding` because it expects that all `ShapeOf` expressions were folded and it basically removes all `Converts` to fp16 and converts model to pure fp16 which is aligned with fp16 in v10 IR

File: tools/mo/openvino/tools/mo/pipeline/common.py [hunk 20]
```diff
@@ -173,16 +173,14 @@ def convert_inputs_of_specific_ops(graph: Graph):
...
     if input_names is None:
         input_names = []
     if meta_info is None:
         meta_info = {}
     graph.strict_mode = False
 
-    # temporary disable new FP16 generation
-    # if not used_by_ir_reader:
-    if True:
+    if convert_types:
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/pipeline/common.py:183]: why do we need to change `Parameter` type here?

File: tools/mo/openvino/tools/mo/main.py [hunk 21]
```diff
@@ -350,7 +348,8 @@ def emit_ir(graph: Graph, argv: argparse.Namespace):
                     mean_data=mean_data,
                     input_names=input_names,
                     meta_info=get_meta_info(argv),
-                    use_temporary_path=True)
+                    use_temporary_path=True,
+                    convert_types=argv.legacy_ir_generation)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:352]: please do not change behavior of deprecated option `legacy_ir_generation`. Please revert

File: tools/mo/openvino/tools/mo/main.py [hunk 22]
```diff
@@ -204,8 +203,7 @@ def raise_ie_not_found():
     except Exception as e:
         raise_ie_not_found()
 
-    # temporary disable new FP16 generation
-    if False and 'data_type' in argv and argv.data_type in ['FP16', 'half']:
+    if not argv.legacy_ir_generation and 'data_type' in argv and argv.data_type in ['FP16', 'half']:
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:206]: please correct. `legacy_ir_generation` is deprecated option

File: tools/mo/openvino/tools/mo/front/tf/bucketize_ext.py [hunk 23]
```diff
@@ -13,6 +14,6 @@ class BucketizeFrontExtractor(FrontExtractorOp):
 
     @classmethod
     def extract(cls, node):
-        boundaries = np.array(node.pb.attr['boundaries'].list.f, dtype=np.float)
+        boundaries = mo_array(node.pb.attr['boundaries'].list.f, dtype=np.float32)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/tf/bucketize_ext.py:17]: float32_array?

File: tools/mo/openvino/tools/mo/front/onnx/group_norm_ext.py [hunk 24]
```diff
@@ -15,8 +16,8 @@ class ExperimentalDetectronGroupNorm(FrontExtractorOp):
     @classmethod
     def extract(cls, node):
         attrs = {
-            'eps': np.array(onnx_attr(node, 'eps', 'f', default=1e-6), dtype=np.float),
-            'num_groups': np.array(onnx_attr(node, 'num_groups', 'i', default=1), dtype=np.int64),
+            'eps': mo_array(onnx_attr(node, 'eps', 'f', default=1e-6), dtype=np.float),
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/front/onnx/group_norm_ext.py:19]: float32_array?

**Review by rkazants — CHANGES_REQUESTED** [review]
please apply comments

File: tools/mo/openvino/tools/mo/front/onnx/group_norm_ext.py (hunk 24 above)
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/onnx/group_norm_ext.py:19]: It is eps, it is float64 array by design

File: tools/mo/openvino/tools/mo/back/compress_quantized_weights.py [hunk 25]
```diff
@@ -103,6 +105,27 @@ def pattern(self):
             ]
         )
 
+    def pattern2(self):
+        return dict(
+            nodes=[
+                ('const', dict(type='Const')),
+                ('const_d', dict()),
+                ('convert', dict(type='Convert')),
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/back/compress_quantized_weights.py:113]: The type of the convert is used in transformation as one of the parameters

File: tools/mo/openvino/tools/mo/pipeline/common.py (hunk 20 above)
Comment (mvafin) [reply tools/mo/openvino/tools/mo/pipeline/common.py:183]: This is done only if convert_types is True. Initially we agreed to not remove code that does old fp16 IR generation, lets do it in separate PR if we want to remove it completely.

File: tools/mo/openvino/tools/mo/front/MatMul_normalizer.py [hunk 26]
```diff
@@ -78,7 +77,7 @@ def find_and_replace_pattern(self, graph: Graph):
             name = node.soft_get('name', node.id)
             node_output_port = node.out_port(0)
             if node.has_valid('alpha') and not math.isclose(node.alpha, 1):
-                mul_alpha = create_op_with_const_inputs(graph, Mul, {1: np.array(node.alpha)},
+                mul_alpha = create_op_with_const_inputs(graph, Mul, {1: mo_array(node.alpha)},
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/MatMul_normalizer.py:80]: We have no way to check the type of the 0 input of MatMul and we never saw fp64 MatMul, so the best type we can guess here is fp32.

File: tools/mo/openvino/tools/mo/front/MatMul_normalizer.py [hunk 27]
```diff
@@ -93,7 +92,7 @@ def find_and_replace_pattern(self, graph: Graph):
                 node.in_port(2).get_connection().set_destination(bias_node.in_port(1))
                 node_output_port.connect(bias_node.in_port(0))
                 if node.has_valid('beta') and not math.isclose(node.beta, 1):
-                    bias_node.insert_op_on_input_port(in_port_idx=1, new_op_class=Mul, value=np.array(node.beta),
+                    bias_node.insert_op_on_input_port(in_port_idx=1, new_op_class=Mul, value=mo_array(node.beta),
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/mo/openvino/tools/mo/front/MatMul_normalizer.py:95]: Same answer

File: tools/mo/openvino/tools/mo/middle/SliceConverter.py [hunk 28]
```diff
@@ -91,7 +91,7 @@ def find_and_replace_pattern(self, graph: Graph):
                 steps = node.in_port(4).data.get_value()
                 assert steps is not None, 'The input with steps is not constant for node {}'.format(node_name)
             else:
-                steps = np.ones([axes.size])
+                steps = np.ones([axes.size], dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (rkazants) [reply tools/mo/openvino/tools/mo/middle/SliceConverter.py:94]: @mvafin, why not int32?

File: tools/mo/openvino/tools/mo/middle/UnsqueezeTileReshapeBlockToInterpolate.py [hunk 29]
```diff
@@ -26,7 +26,7 @@ class UnsqueezeTileReshapeBlockToInterpolate(MiddleReplacementPattern):
     one after another, performing unsqueeze-ing over different dimensions, effectively performing interpolation over
     several dimensions.
 
-    These sequences are merged in the 'optimizer/openvino/tools/mo/middle/InterpolateSequenceToInterpolate.py' transformation
+    These sequences are merged in the 'optimizer/extensions/middle/InterpolateSequenceToInterpolate.py' transformation
```
[the remark this answers is not in the document]
Comment (rkazants) [reply tools/mo/openvino/tools/mo/middle/UnsqueezeTileReshapeBlockToInterpolate.py:29]: please correct a path: mo/middle/InterpolateSequenceToInterpolate.py

File: tools/mo/openvino/tools/mo/ops/pooling.py [hunk 30]
```diff
@@ -132,7 +133,7 @@ def pool_infer(node: Node):
             node.window[node.spatial_dims] = input_spatial_shape
 
         if not node.has_valid('dilation'):
-            node['dilation'] = np.ones(len(input_shape))
+            node['dilation'] = np.ones(len(input_shape), dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (rkazants) [reply tools/mo/openvino/tools/mo/ops/pooling.py:136]: @mvafin, please comment

File: tools/mo/openvino/tools/mo/ops/convolution.py [hunk 31]
```diff
@@ -287,3 +290,18 @@ def reverse_infer(node: Node):
...
+        in_type_1 = node.in_port(1).get_data_type()
+        in_node_1 = node.in_port(1).get_source().node
+        # in case of input values data type mismatch we try to change the type of the constant to match the type of
+        # input at index 0.
+        if in_type_1 in [np.float16, np.float32, np.float64] and in_type_0 != in_type_1 and in_node_1.op == 'Const':
+            in_node_1 = node.in_port(1).get_source().node
+            log.error("Changing Const node '{}' data type from {} to {} for Convolution operation".format(
+                in_node_1.soft_get('name', in_node_1.id), in_type_1, in_type_0),
+                extra={'is_warning': True})
+            convert_const_node_value_type(in_node_1, in_type_0)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/ops/convolution.py:306]: What cases exactly does this logic cover? input and weight types?

**Review by rkazants — COMMENTED** [review]
In general, it looks great. Remained some comments not so relevant now if we get normal DLB results. Let's anticipate them and merge if it is fine.

**Review by rkazants — APPROVED** [review]
